pkg/execution/state/redis_state/queue_processor.go

```diff
@@ -228,6 +237,171 @@ LOOP:
+	leaseTick := time.NewTicker(ShardLeaseTime / 2)
+
+	// records whether we're leasing
+	var leasing int32
+
+	for {
+		select {
+		case <-ctx.Done():
+			// TODO: Remove leases immediately from backing store.
+			scanTick.Stop()
+			leaseTick.Stop()
+			return
+		case <-scanTick.C:
+			go func() {
+				if !atomic.CompareAndSwapInt32(&leasing, 0, 1) {
+					// Only one lease can occur at once.
+					return
+				}
+
+				// Always reset the leasing op to zero, allowing us to lease again.
+				defer func() { atomic.StoreInt32(&leasing, 0) }()
+
+				// Retry claiming leases until all shards have been taken.  All operations
+				// must succeed, even if it leaves us spinning.  Note that scanShards filters
+				// out unnecessary leases and shards that have already been leased.
+				retry := true
+				for retry {
+					var err error
+					retry, err = q.scanShards(ctx)
+					if err != nil {
+						q.logger.Error().Err(err).Msg("error scanning and leasing shards")
+						return
+					}
+				}
+			}()
+		case <-leaseTick.C:
+			for _, s := range q.shardLeases {
+				// Attempt to lease all ASAP, even if the backing store is single threaded.
+				go func(ls leasedShard) {
+					nextLeaseID, err := q.renewShardLease(ctx, &ls.Shard, ShardLeaseTime, ls.Lease)
+					if err != nil {
+						q.logger.Error().Err(err).Msg("error renewing shard lease")
+						return
+					}
+					// Update the lease ID so that we have this stored appropriately for
+					// the next renewal.
+					q.addLeasedShard(ctx, &ls.Shard, *nextLeaseID)
+				}(s)
+			}
+		}
+	}
+}
+
+func (q *queue) scanShards(ctx context.Context) (retry bool, err error) {
+	// TODO: Make instances of *queue register worker information when calling
+	//       Run().
+	//       Fetch this information, and correctly assign workers to shard maps
+	//       based on the distribution of items in the queue here.  This lets
+	//       us oversubscribe appropriately.
+
+	shardMap, err := q.getShards(ctx)
+	if err != nil {
+		q.logger.Error().Err(err).Msg("error fetching shards")
+		return
+	}
+	shards, err := q.filterShards(ctx, shardMap)
+	if err != nil {
+		q.logger.Error().Err(err).Msg("error filtering shards")
+		return
+	}
+	if len(shards) == 0 {
+		return
+	}
+
+	for _, shard := range shards {
+		leaseID, err := q.leaseShard(ctx, shard, ShardLeaseTime, len(shard.Leases))
+		if err == nil {
+			q.addLeasedShard(ctx, shard, *leaseID)
+		}
+		switch err {
+		case errShardNotFound:
+			// This is okay;  the shard was removed when trying to lease
+			continue
+		case errShardIndexLeased:
+			// This is okay;  another worker grabbed the lease.  No need to retry
+			// as another worker grabbed this.
+			continue
+		case errShardIndexInvalid:
+			// A lease expired while trying to lease — try again.
+			retry = true
+		default:
+			return true, err
+		}
+	}
+
+	return retry, nil
+}
+
+func (q *queue) addLeasedShard(ctx context.Context, shard *QueueShard, lease ulid.ULID) {
+	for i, n := range q.shardLeases {
```

Comment:
do these operations need locking?
I guess in other words, will tick in `scanShards` ever cause >1 go routine to run?

Reply:
Correct, let me add that.  Thanks.